File: app/src/main/java/com/telolahy/towerofhanoi/object/LevelCompleteWindow.java
package com.telolahy.towerofhanoi.object;

import com.telolahy.towerofhanoi.Constants;
import com.telolahy.towerofhanoi.R;
import com.telolahy.towerofhanoi.manager.GameManager;
import com.telolahy.towerofhanoi.manager.ResourcesManager;

import org.andengine.engine.camera.Camera;
import org.andengine.entity.scene.Scene;
import org.andengine.entity.sprite.Sprite;
import org.andengine.entity.sprite.TiledSprite;
import org.andengine.entity.text.Text;
import org.andengine.input.touch.TouchEvent;
import org.andengine.opengl.vbo.VertexBufferObjectManager;

/**
 * Created by stephanohuguestelolahy on 11/8/14.
 */
public class LevelCompleteWindow extends Sprite {


    public interface LevelCompleteWindowListener {

        void levelCompleteWindowNextButtonClicked();

        void levelCompleteWindowReplayButtonClicked();
    }

    private TiledSprite mStar1;
    private TiledSprite mStar2;
    private TiledSprite mStar3;
    private LevelCompleteWindowListener mListener;

    public enum StarsCount {
        ONE,
        TWO,
        THREE
    }

    public LevelCompleteWindow(VertexBufferObjectManager pSpriteVertexBufferObject, Scene scene, LevelCompleteWindowListener listener) {

        super(400, 240, ResourcesManager.getInstance().gameWindowRegion, pSpriteVertexBufferObject);
        mListener = listener;
        attachStars(pSpriteVertexBufferObject, scene);
    }

    private void attachStars(VertexBufferObjectManager pSpriteVertexBufferObject, Scene scene) {

        ResourcesManager resourcesManager = ResourcesManager.getInstance();

        String text = resourcesManager.activity.getResources().getString(R.string.level_completed);
        boolean isOnLastLevel = GameManager.getInstance().isOnLastLevel();
        if (isOnLastLevel) {
            text = resourcesManager.activity.getResources().getString(R.string.last_level_completed);
        }

        attachChild(new Text(Constants.CAMERA_WIDTH / 2, Constants.CAMERA_HEIGHT - 100, resourcesManager.mainFont.font, text, resourcesManager.vertexBufferObjectManager));

        mStar1 = new TiledSprite(275, 260, resourcesManager.gameCompleteStarsRegion, pSpriteVertexBufferObject);
        mStar2 = new TiledSprite(400, 260, resourcesManager.gameCompleteStarsRegion, pSpriteVertexBufferObject);
        mStar3 = new TiledSprite(525, 260, resourcesManager.gameCompleteStarsRegion, pSpriteVertexBufferObject);
        attachChild(mStar1);
        attachChild(mStar2);
        attachChild(mStar3);

        Sprite retryButton = new Sprite(260, 120, resourcesManager.gameCompleteRetryRegion, pSpriteVertexBufferObject) {
            @Override
            public boolean onAreaTouched(TouchEvent pSceneTouchEvent, float pTouchAreaLocalX, float pTouchAreaLocalY) {
                mListener.levelCompleteWindowReplayButtonClicked();
                return true;
            }
        };
        attachChild(retryButton);
        scene.registerTouchArea(retryButton);


        if (isOnLastLevel) {

            retryButton.setPosition(400, retryButton.getY());
            return;
        }

        Sprite nextButton = new Sprite(540, 120, resourcesManager.gameCompleteNextRegion, pSpriteVertexBufferObject) {
            @Override
            public boolean onAreaTouched(TouchEvent pSceneTouchEvent, float pTouchAreaLocalX, float pTouchAreaLocalY) {
                mListener.levelCompleteWindowNextButtonClicked();
                return true;
            }
        };
        attachChild(nextButton);
        scene.registerTouchArea(nextButton);
    }

    /**
     * Change star`s tile index, depends on stars count.
     *
     * @param starsCount
     */
    public void display(StarsCount starsCount, Scene scene, Camera camera) {
        // Change stars tile index, based on stars count (1-3)
        switch (starsCount) {
            case ONE:
                mStar1.setCurrentTileIndex(0);
                mStar2.setCurrentTileIndex(1);
                mStar3.setCurrentTileIndex(1);
                break;
            case TWO:
                mStar1.setCurrentTileIndex(0);
                mStar2.setCurrentTileIndex(0);
                mStar3.setCurrentTileIndex(1);
                break;
            case THREE:
                mStar1.setCurrentTileIndex(0);
                mStar2.setCurrentTileIndex(0);
                mStar3.setCurrentTileIndex(0);
                break;
        }


        // Hide HUD
        camera.getHUD().setVisible(false);

        // Disable camera chase entity
        camera.setChaseEntity(null);

        // Attach our level complete panel in the middle of camera
        setPosition(camera.getCenterX(), camera.getCenterY());
        scene.attachChild(this);
    }
}

File: app/src/main/java/com/telolahy/towerofhanoi/object/LevelCompleteWindow.kt
package com.telolahy.towerofhanoi.`object`

import com.telolahy.towerofhanoi.Constants
import com.telolahy.towerofhanoi.R
import com.telolahy.towerofhanoi.manager.GameManager
import com.telolahy.towerofhanoi.manager.ResourcesManager
import org.andengine.engine.camera.Camera
import org.andengine.entity.scene.Scene
import org.andengine.entity.sprite.Sprite
import org.andengine.entity.sprite.TiledSprite
import org.andengine.entity.text.Text
import org.andengine.input.touch.TouchEvent
import org.andengine.opengl.vbo.VertexBufferObjectManager

/**
 * Created by stephanohuguestelolahy on 11/8/14.
 */
class LevelCompleteWindow(pSpriteVertexBufferObject: VertexBufferObjectManager, scene: Scene, private val mListener: LevelCompleteWindowListener) : Sprite(Constants.CAMERA_WIDTH / 2, Constants.CAMERA_HEIGHT / 2, ResourcesManager.instance.gameCompleteWindowTexture.textureRegion, pSpriteVertexBufferObject) {

  interface LevelCompleteWindowListener {
    fun levelCompleteWindowNextButtonClicked()
    fun levelCompleteWindowReplayButtonClicked()
  }

  private var mStar1: TiledSprite? = null
  private var mStar2: TiledSprite? = null
  private var mStar3: TiledSprite? = null

  enum class StarsCount {
    ONE, TWO, THREE
  }

  private fun attachStars(pSpriteVertexBufferObject: VertexBufferObjectManager, scene: Scene) {
    val resourcesManager = ResourcesManager.instance
    var text = resourcesManager.activity.resources.getString(R.string.level_completed)
    val isOnLastLevel = GameManager.instance.isOnLastLevel
    if (isOnLastLevel) {
      text = resourcesManager.activity.resources.getString(R.string.last_level_completed)
    }
    attachChild(Text(Constants.CAMERA_WIDTH / 2, Constants.CAMERA_HEIGHT - 100, resourcesManager.mainFont.font, text, resourcesManager.vertexBufferObjectManager))
    mStar1 = TiledSprite(275f, 260f, resourcesManager.gameCompleteStarsTexture.textureRegion, pSpriteVertexBufferObject)
    mStar2 = TiledSprite(400f, 260f, resourcesManager.gameCompleteStarsTexture.textureRegion, pSpriteVertexBufferObject)
    mStar3 = TiledSprite(525f, 260f, resourcesManager.gameCompleteStarsTexture.textureRegion, pSpriteVertexBufferObject)
    attachChild(mStar1)
    attachChild(mStar2)
    attachChild(mStar3)
    val retryButton: Sprite = object : Sprite(260f, 120f, resourcesManager.gameCompleteRetryTexture.textureRegion, pSpriteVertexBufferObject) {
      override fun onAreaTouched(pSceneTouchEvent: TouchEvent, pTouchAreaLocalX: Float, pTouchAreaLocalY: Float): Boolean {
        mListener.levelCompleteWindowReplayButtonClicked()
        return true
      }
    }
    attachChild(retryButton)
    scene.registerTouchArea(retryButton)
    if (isOnLastLevel) {
      retryButton.setPosition(400f, retryButton.y)
      return
    }
    val nextButton: Sprite = object : Sprite(540f, 120f, resourcesManager.gameCompleteNextTexture.textureRegion, pSpriteVertexBufferObject) {
      override fun onAreaTouched(pSceneTouchEvent: TouchEvent, pTouchAreaLocalX: Float, pTouchAreaLocalY: Float): Boolean {
        mListener.levelCompleteWindowNextButtonClicked()
        return true
      }
    }
    attachChild(nextButton)
    scene.registerTouchArea(nextButton)
  }

  /**
   * Change star`s tile index, depends on stars count.
   *
   * @param starsCount
   */
  fun display(starsCount: StarsCount?, scene: Scene, camera: Camera) {
    // Change stars tile index, based on stars count (1-3)
    when (starsCount) {
      StarsCount.ONE -> {
        mStar1?.currentTileIndex = 0
        mStar2?.currentTileIndex = 1
        mStar3?.currentTileIndex = 1
      }
      StarsCount.TWO -> {
        mStar1?.currentTileIndex = 0
        mStar2?.currentTileIndex = 0
        mStar3?.currentTileIndex = 1
      }
      StarsCount.THREE -> {
        mStar1?.currentTileIndex = 0
        mStar2?.currentTileIndex = 0
        mStar3?.currentTileIndex = 0
      }

        null -> TODO()
    }


    // Hide HUD
    camera.hud.isVisible = false

    // Disable camera chase entity
    camera.setChaseEntity(null)

    // Attach our level complete panel in the middle of camera
    setPosition(camera.centerX, camera.centerY)
    scene.attachChild(this)
  }

  init {
    attachStars(pSpriteVertexBufferObject, scene)
  }
}